packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,275 @@
+  /**
+   * a ReactNode or a function that returns the content of the dialog.
+   */
+  children?:
+    | ReactNode
+    | (({ onClose }: { onClose?: DialogProps['onClose'] }) => ReactNode);
+  /**
+   * Text label for the close button for screen readers.
+   * Important for accessibility.
+   */
+  closeButtonLabel: string;
+  /**
+   * Use the `immersive` variant to focus the user's attention on the dialog content.
+   * default: 'contextual'
+   * */
+  variant?: 'contextual' | 'immersive';
+  /**
+   * Prevent users from closing the modal by clicking/tapping the overlay or
+   * pressing the escape key. Default `false`.
+   */
+  preventClose?: boolean;
+  /**
+   * Enables focusing a particular element in the dialog content and override default behavior
+   */
+  initialFocusRef?: RefObject<HTMLElement>;
+}
+
+export const animationDuration = 300;
+
+export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
+  (
+    {
+      open,
+      onClose,
+      closeButtonLabel,
+      variant = 'contextual',
+      children,
+      className,
+      preventClose,
+      initialFocusRef,
+      ...props
+    },
+    ref,
+  ) => {
+    const dialogRef = useRef<HTMLDialogElement>(null);
+    const lastFocusedElementRef = useRef<HTMLElement | null>(null);
+    // TODO add translated default close label
+    if (process.env.NODE_ENV !== 'production') {
+      if (!isSufficientlyLabelled(closeButtonLabel)) {
+        throw new AccessibilityError(
+          'Dialog',
+          'The `closeButtonLabel` prop is missing or invalid.',
+        );
+      }
+    }
+    const hasNativeDialog = hasNativeDialogSupport();
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+      if (open && dialogElement) {
+        setTimeout(() => {
+          if (initialFocusRef) {
+            initialFocusRef?.current?.focus();
+          } else {
+            getFirstFocusableElement(dialogElement).focus();
+          }
+        }, animationDuration);
+      }
+    }, [open]);
+
+    const handleDialogClose = useCallback(() => {
+      const dialogElement = dialogRef.current;
+      if (!dialogElement) {
+        return;
+      }
+      // restore focus to the last focused element
+      if (lastFocusedElementRef.current) {
+        setTimeout(() => lastFocusedElementRef.current?.focus(), 300);
+      }
+      // restore scroll tp page
+      document.documentElement.style.overflowY = 'unset';
+      // trigger close animation
+      dialogElement.classList.remove(classes.show);
+      if (!hasNativeDialog) {
+        (dialogElement.nextSibling as HTMLDivElement).classList.remove(
+          classes['backdrop-visible'],
+        );
+      }
+      // trigger dialog close after animation
+      setTimeout(() => {
+        if (dialogElement.open) {
+          dialogElement.close();
+        }
+      }, animationDuration);
+    }, []);
+
+    function onPolyfillDialogKeydown(event: KeyboardEvent) {
+      if (isEscape(event) && preventClose) {
```

Comment:
Minuscule performance improvement:

```suggestion
      if (preventClose && isEscape(event)) {
```

Reply:
Actually, I noticed that the event listener is only attached below when `preventClose` is truthy, so the check here seems redundant.